src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,567 @@
+   * Throws exception if column index is out of bounds.
+   *
+   * @param index column index
+   * @return matrix of draws across all chains
+   */
+  Eigen::MatrixXd samples(const int index) const {
+    Eigen::MatrixXd result(num_samples(), chains_.size());
+    if (index < 0 || index >= param_names_.size()) {
+      std::stringstream ss;
+      ss << "Bad index " << index << ", should be between 0 and "
+         << (param_names_.size() - 1);
+      throw std::invalid_argument(ss.str());
+    }
+    for (int i = 0; i < chains_.size(); ++i) {
+      result.col(i) = chains_[i].col(index);
+    }
+    return result;
+  }
+
+  /**
+   * Assemble samples (draws) from specified parameter name across all chains
+   * into a matrix of samples X chain.
+   * Throws exception if parameter name is not found.
+   *
+   * @param name parameter name
+   * @return matrix of draws across all chains
+   */
+  Eigen::MatrixXd samples(const std::string& name) const {
+    return samples(index(name));
+  }
+
+  /**
+   * Compute mean value for specified parameter across all chains.
+   *
+   * @param index parameter index
+   * @return mean parameter value
+   */
+  double mean(const int index) const { return samples(index).mean(); }
+
+  /**
+   * Compute mean value for specified parameter across all chains.
+   *
+   * @param name parameter name
+   * @return mean parameter value
+   */
+  double mean(const std::string& name) const { return mean(index(name)); }
+
+  /**
+   * Compute sample variance for specified parameter across all chains.
+   * 1 / (N - 1) * sum((theta_n - mean(theta))^2)
+   *
+   * @param index parameter index
+   * @return sample variance
+   */
+  double variance(const int index) const {
+    Eigen::MatrixXd draws = samples(index);
+    return (draws.array() - draws.mean()).square().sum() / (draws.size() - 1);
+  }
+
+  /**
+   * Compute sample variance for specified parameter across all chains.
+   * 1 / (N - 1) * sum((theta_n - mean(theta))^2)
+   *
+   * @param name parameter name
+   * @return sample variance
+   */
+  double variance(const std::string& name) const {
+    return variance(index(name));
+  }
+
+  /**
+   * Compute standard deviation for specified parameter across all chains.
+   *
+   * @param index parameter index
+   * @return sample sd
+   */
+  double sd(const int index) const { return std::sqrt(variance(index)); }
+
+  /**
+   * Compute standard deviation for specified parameter across all chains.
+   *
+   * @param name parameter name
+   * @return sample sd
+   */
+  double sd(const std::string& name) const { return sd(index(name)); }
+
+  /**
+   * Compute median value of specified parameter across all chains.
+   *
+   * @param index parameter index
+   * @return median
+   */
+  double median(const int index) const {
+    Eigen::MatrixXd draws = samples(index);
+    std::vector<double> sorted(draws.data(), draws.data() + draws.size());
+    std::sort(sorted.begin(), sorted.end());
```

Comment:
Optional: I think there is a faster way than sorting. `std::nth_element` looks close, but not perfect

Reply:
implemented.  note, other than for median, we'll likely be calling `quantiles`, and use of std::sort is unavoidable.

Reply:
we avoid it in the existing `chains.hpp` which uses `boost::accumulators::quantile`. is there a reason to change?

Reply:
for median and `quantile`, std::nth_element is fast.
for `quantiles`, is sorting the entire array once better than iterating on a bunch of partial sorts?

the code in `chains.hpp` `quantiles` function also is just a big loop over the array of probs passed in.

Reply:
Assuming the boost quantile code is one of the O(n) implementations, for p quantiles that will be O(np)

Sorting is O(n log(n)), so if p > log(n), I would expect the boost version with a loop to be faster, but this is where constant factors would start to mater for small n, p

Reply:
my other concern, w/r/t quantiles, is to avoid the copy that's required in setting up the nth_element finding array.
also, N is usually 4000, so logN is 12, and the default quantiles call has 3 args.  so a difference of 9, plus whatever constant, plus 2 extra copy operations.   I would prefer to leave it as is.

Reply:
This sort also requires a copy right?

By "leave it as is" does that mean the boost implementation?

Reply:
for `quantiles`, using std::sort requires a copy, but then you can use the sorted array to get all requested quantiles.
added `quantile`, which uses `nth_element`, so one copy but less sorting.  `median` calls `quantile`.

Reply:
Does this actually compute the same quantiles as the current code? There are a fair number of subtleties, cf https://github.com/stan-dev/cmdstanpy/issues/762#issuecomment-2150061089

Reply:
this computes the same quantiles as the implementation in R's `posterior` package, which is what CmdStanR uses.  

latest commit is using `nth_element` everywhere.   I don't think we need the boost accumulators includes anymore.

Reply:
https://github.com/stan-dev/stan/blob/021ac040c0d18dd97c5311b033b5e58c21502731/src/test/unit/mcmc/chainset_test.cpp#L179

Reply:
Okay. Probably still worth noting to users that this will be different than the numbers they've been seeing before